File: src/world/physical_object.go
use maths;

// Gravity is the gravity acceleration constant (m/s/s)
const Gravity = -9.81

// PhysicalObject is an object with physics, position,
// velocity, and mass
struct PhysicalObject {
	frozen bool // if true, the PhysicalObject will not move

	velocity     maths.Vec3
	acceleration maths.Vec3
	mass         f32 // In kilograms.

	onGround    bool
	pushingWall bool
	atCeiling   bool

	wasOnGround    bool
	wasPushingWall bool
	wasAtCeiling   bool

	OnGroundhit  fn()
	OnPush       fn()
	OnCeilinghit fn()

	hitbox &maths.AABC // hitbox for collision calculation, but not kill calculation
}

// Position returns the PhysicalObject's position
fn Position(&self) maths.Vec3 {
	return p.hitbox.CenterPos
}

fn AddPosition(&self, v2 maths.Vec3) {
	p.hitbox.CenterPos.Add(v2)
}

// SetPosition modifies the position of the PhysicalObject.
fn SetPosition(&self, pos maths.Vec3) {
	p.hitbox.CenterPos = pos
}

// Physics calculates physics on the PhysicalObject p.
fn Physics(&self, delta f32) {
	// no physics if p is frozen
	if p.frozen {
		return
	}

	// gravity only applies if the PhysicalObject is not on
	// ground
	if !p.onGround {
		p.acceleration.Y += Gravity
	}

	p.acceleration.Scale(delta) // converts to velocity
	p.velocity.Add(p.acceleration)

	p.velocity.Scale(delta) // converts to displacement
	p.hitbox.CenterPos.Add(p.velocity)

	// reset acceleration
	p.acceleration.X = 0
	p.acceleration.Y = 0
	p.acceleration.Z = 0
}

// ApplyForce applies a force, in newtons, to the
// PhysicalObject. This is the only way to move a
// PhysicalObject in the game; velocity and acceleration are
// not publicly accessible.
fn ApplyForce(&self, newtons maths.Vec3) {
	p.acceleration.X += newtons.X / p.mass
	p.acceleration.Y += newtons.Y / p.mass
	p.acceleration.Z += newtons.Z / p.mass
}

// StopMotion immediately stops the motion of the
// PhysicalObject. Velocity and acceleration are set to
// zero.
fn StopMotion(&self) {
	p.velocity.X = 0
	p.velocity.Y = 0
	p.velocity.Z = 0
	p.acceleration.X = 0
	p.acceleration.Y = 0
	p.acceleration.Z = 0
}

// CollidesWith returns whether or not the Collider
// collides with another Collider
fn CollidesWith(&self, other &PhysicalObject) bool {
	return p.hitbox.CollidesWith(other.hitbox)
}

// FixCollision fixes a collision between two
// PhysicalObjects. If both objects are actively subject to
// forces, momentum will take effect on both PhysicalObjects
// and force will be applied to both of  them
fn FixCollision(&self, other &PhysicalObject) {
	if p.frozen || !p.CollidesWith(other) {
		return
	}

	// fix collisions
	switch {
	case !other.frozen:
		// fix both
		firstBreach = calculateBreach(p, other)
		secondBreach = calculateBreach(other, p)

		firstBreach.Scale(0.5)
		secondBreach.Scale(0.5)

		fix(p, firstBreach)
		fix(other, secondBreach)

		applyMomentum(p, other)
	case other.frozen:
		breach = calculateBreach(p, other)

		// fix p
		fix(p, breach)

		// now, we need to determine on which side of
		// `other` p is on. if it's on the top or bottom,
		// velocity stops on the y axis. if left or right, x
		// axis. first, re-calculate the breach now that
		// we've fixed the objects:
		breach = calculateBreach(p, other)

		// smallest breach determines which side the object is on
		minBreach = math.Min(breach.X), math.Min(f64(breach.Y), f64(breach.Z)) as f64 as f32
		switch minBreach {
		case breach.X:
			p.velocity.X = 0
		case breach.Y:
			p.velocity.Y = 0
		case breach.Z:
			p.velocity.Z = 0
		}
	}
}

fn calculateBreach(moving, static &PhysicalObject) (breach maths.Vec3) {
	// breach really depends on which direction the moving
	// PhysicalObject is travelling

	// calculate X
	switch {
	case moving.velocity.X > 0:
		breach.X = moving.hitbox.CenterPos.X + moving.hitbox.HalfSize.X - (static.hitbox.CenterPos.X - static.hitbox.HalfSize.X)
	case moving.velocity.X < 0:
		breach.X = moving.hitbox.CenterPos.X - moving.hitbox.HalfSize.X - (static.hitbox.CenterPos.X + static.hitbox.HalfSize.X)
	}

	// calculate Y
	switch {
	case moving.velocity.Y > 0:
		breach.Y = moving.hitbox.CenterPos.Y + moving.hitbox.HalfSize.Y - (static.hitbox.CenterPos.Y - static.hitbox.HalfSize.Y)
	case moving.velocity.Y < 0:
		breach.Y = moving.hitbox.CenterPos.Y - moving.hitbox.HalfSize.Y - (static.hitbox.CenterPos.Y + static.hitbox.HalfSize.Y)
	}

	// calculate Z
	switch {
	case moving.velocity.Z > 0:
		breach.Z = moving.hitbox.CenterPos.Z + moving.hitbox.HalfSize.Z - (static.hitbox.CenterPos.Z - static.hitbox.HalfSize.Z)
	case moving.velocity.Z < 0:
		breach.Z = moving.hitbox.CenterPos.Z - moving.hitbox.HalfSize.Z - (static.hitbox.CenterPos.Z + static.hitbox.HalfSize.Z)
	}
	return
}

// vim: foldmethod=syntax

File: src/world/physical_object.rs
use crate::maths;

/// The gravity acceleration constant (m/s/s)
const GRAVITY: f32 = -9.81;

/// PhysicalObject is an object with physics, position,
/// velocity, and mass
struct PhysicalObject {
    /// if true, the PhysicalObject will not move
    frozen: bool,

    velocity: Vector3,
    acceleration: Vector3,

    /// In kilograms.
    mass: f32,

    on_ground: bool,
    pushing_wall: bool,
    at_ceiling: bool,

    was_on_ground: bool,
    was_pushing_wall: bool,
    was_at_ceiling: bool,

    hitbox: maths::AABC, // hitbox for collision calculation, but not kill calculation
}

impl PhysicalObject {
    /// Returns the PhysicalObject's position
    fn position(&self) -> Vector3 {
        self.hitbox.center_pos;
    }

    fn add_position(&self, v2: Vector3) {
        self.hitbox.center_pos.Add(v2)
    }

    /// Modifies the position of the PhysicalObject.
    fn set_position(&self, pos: Vector3) {
        self.hitbox.center_pos = pos
    }

    /// Calculates physics on the PhysicalObject.
    fn physics(&self, delta: f32) {
        // no physics if p is frozen
        if self.frozen {
            return;
        }

        // gravity only applies if the PhysicalObject is not on
        // ground
        if !self.on_ground {
            self.acceleration.y += gravity;
        }

        self.acceleration.scale(delta); // converts to velocity
        self.velocity.add(self.acceleration);

        self.velocity.scale(delta); // converts to displacement
        self.hitbox.center_pos.add(self.velocity);

        // reset acceleration
        self.acceleration.x = 0;
        self.acceleration.y = 0;
        self.acceleration.z = 0;
    }

    /// Applies a force, in newtons, to the PhysicalObject. This is the only way to move a
    /// PhysicalObject in the game; velocity and acceleration are not publicly accessible.
    fn apply_force(&self, newtons: Vector3) {
        self.acceleration.x += newtons.x / self.mass;
        self.acceleration.y += newtons.y / self.mass;
        self.acceleration.z += newtons.z / self.mass;
    }

    /// Immediately stops the motion of the PhysicalObject. Velocity and acceleration are set to
    /// zero.
    fn stop_motion(&self) {
        self.velocity.x = 0;
        self.velocity.y = 0;
        self.velocity.z = 0;
        self.acceleration.x = 0;
        self.acceleration.y = 0;
        self.acceleration.z = 0;
    }

    /// Returns whether or not the PhysicalObjects collide.
    fn collides_with(&self, other: &PhysicalObject) -> bool {
        self.hitbox.collides_with(other.hitbox)
    }

    /// FixCollision fixes a collision between two PhysicalObjects. If both objects are actively
    /// subject to forces, momentum will take effect on both PhysicalObjects and force will be
    /// applied to both of  them
    fn fix_collision(&self, other: &PhysicalObject) {
        if self.frozen || !self.collides_with(other) {
            return;
        }

        // fix collisions
        if !other.frozen {
            // fix both
            first_breach = calculate_breach(p, other);
            second_breach = calculate_breach(other, p);

            first_breach.scale(0.5);
            second_breach.scale(0.5);

            fix(p, first_breach);
            fix(other, second_breach);

            apply_momentum(p, other);
        } else if other.frozen {
            breach = calculate_breach(p, other);

            // fix p
            fix(p, breach);

            // now, we need to determine on which side of
            // `other` p is on. if it's on the top or bottom,
            // velocity stops on the y axis. if left or right, x
            // axis. first, re-calculate the breach now that
            // we've fixed the objects:
            breach = calculate_breach(p, other);

            // smallest breach determines which side the object is on
            min_breach = breach.x.min(breach.y.min(breach.z));

            if breach.x == minBreach {
                self.velocity.x = 0;
            } else if breach.y == minBreach {
                self.velocity.y = 0;
            } else if breach.z == minBreach {
                self.velocity.z = 0;
            }
        }
    }
}

/// Returns a breach.
fn calculate_breach(moving: &PhysicalObject, still: &PhysicalObject) -> Vector3 {
    // breach really depends on which direction the moving
    // PhysicalObject is travelling

    // calculate x
    if moving.velocity.x > 0 {
        breach.x = moving.hitbox.center_pos.x + moving.hitbox.half_size.x
            - (still.hitbox.center_pos.x - still.hitbox.half_size.x);
    } else if moving.velocity.x < 0 {
        breach.x = moving.hitbox.center_pos.x
            - moving.hitbox.half_size.x
            - (still.hitbox.center_pos.x + still.hitbox.half_size.x);
    }

    // calculate y
    if moving.velocity.y > 0 {
        breach.y = moving.hitbox.center_pos.y + moving.hitbox.half_size.y
            - (still.hitbox.center_pos.y - still.hitbox.half_size.y);
    } else if moving.velocity.y < 0 {
        breach.y = moving.hitbox.center_pos.y
            - moving.hitbox.half_size.y
            - (still.hitbox.center_pos.y + still.hitbox.half_size.y);
    }

    // calculate z
    if moving.velocity.z > 0 {
        breach.z = moving.hitbox.center_pos.z + moving.hitbox.half_size.z
            - (still.hitbox.center_pos.z - still.hitbox.half_size.z);
    } else if moving.velocity.z < 0 {
        breach.z = moving.hitbox.center_pos.z
            - moving.hitbox.half_size.z
            - (still.hitbox.center_pos.z + still.hitbox.half_size.z);
    }

    breach
}

// vim: foldmethod=syntax
